File: pkg/yopass/redis.go
package yopass

import (
	"time"

	"github.com/go-redis/redis/v7"
)

// NewRedis returns a new Redis database client
func NewRedis(url string) (Database, error) {
	options, err := redis.ParseURL(url)
	if err != nil {
		return nil, err
	}
	client := redis.NewClient(options)
	return &Redis{client}, nil
}

// Redis client
type Redis struct {
	client *redis.Client
}

// Get key from Redis
func (r *Redis) Get(key string) (string, error) {
	v, err := r.client.Get(key).Result()
	if err != nil {
		return "", err
	}
	return v, nil
}

// Put key to Redis
func (r *Redis) Put(key, value string, expiration int32) error {
	return r.client.Set(
		key,
		value,
		time.Duration(expiration)*time.Second,
	).Err()
}

// Delete key from Redis
func (r *Redis) Delete(key string) error {
	return r.client.Del(key).Err()
}

File: next/src/database/redis.ts
import { Database, GetResponse, StoreRequest, StoreResponse } from './database';
import { createClient, RedisClientType } from 'redis';

type RedisOptions = {
  url?: string;
};

export class Redis implements Database {
  private constructor(private readonly client: RedisClientType<any, any>) {}

  static async create(options: RedisOptions): Promise<Redis> {
    const client = createClient({ url: options.url });
    await client.connect();
    return new Redis(client);
  }

  async get(options: { key: string }): Promise<GetResponse> {
    const result = await this.client.get(options.key);

    if (!result) {
      throw Error('Secret not found');
    }
    const data: GetResponse = JSON.parse(result);
    return data;
  }

  async store(options: StoreRequest): Promise<StoreResponse> {
    const { ttl, oneTime, key } = options;
    await this.client.set(
      key,
      JSON.stringify({
        message: options.secret,
        ttl,
        oneTime,
      }),
      { EX: ttl },
    );
    return { key };
  }

  async delete(options: { key: string }): Promise<void> {
    await this.client.del(options.key);
  }
}
